src/server/lib/okta/register.ts

```diff
@@ -71,27 +151,79 @@ export const register = async (email: string): Promise<UserResponse> => {
  *
  * We read the state of the user and send the appropriate email.
  *
- * TODO: Implement recovery for Status.ACTIVE || Status.RECOVERY || Status.PASSWORD_EXPIRED
- *
  * @param {string} email
  * @returns {Promise<UserResponse>} Promise that resolves to the user object
  */
-export const resendRegistrationEmail = async (email: string) => {
+export const resendRegistrationEmail = async (
+  email: string,
+): Promise<UserResponse> => {
   const user: UserResponse = await getUser(email);
   const { status } = user;
   switch (status) {
-    case Status.STAGED: {
-      await activateUser(email);
-      break;
-    }
+    case Status.STAGED:
     case Status.PROVISIONED: {
-      await reactivateUser(email);
-      break;
+      /* Given I'm a STAGED or PROVISIONED user
+       *    When I ask for my email to be resent
+       *    Then Gateway should ask Okta for my activation token
+       *    And I should be sent a set password email with the activation
+       *    token through Gateway
+       */
+      const tokenResponse = await reactivateUser(user.id, false);
+      if (tokenResponse?.token.length) {
+        const emailIsSent = await sendAccountExistsEmail({
+          to: user.profile.email,
+          activationToken: tokenResponse.token,
+        });
+        if (emailIsSent) {
+          return user;
+        } else {
+          throw new OktaError({
+            message: `Okta user activation failed: Failed to send email`,
+          });
+        }
+      } else {
+        throw new OktaError({
+          message: `Okta registration failed: missing re-activation token`,
+        });
+      }
```

Comment:
I'm generally more of a fan of guard clauses, as it makes errors and when you throw errors a bit easier to read

So something like this, instead of nesting if/else statements:

```suggestion
if (!tokenResponse?.token.length) {
  throw new OktaError({
    message: `Okta registration failed: missing re-activation token`,
  });
}

const emailIsSent = await sendAccountExistsEmail({
  to: user.profile.email,
  activationToken: tokenResponse.token,
});

if (!emailIsSent) {
  throw new OktaError({
    message: `Okta user activation failed: Failed to send email`,
  });
}

return user;
```

Reply:
can implement a similar thing below too for the other cases too above and below

Reply:
Yeah, this is great, I prefer this too!